### PR TITLE
Change Exit contract to specific implementation

### PR DESCRIPTION
**Pull request by Akira-Taniguchi**
# Description

・Create an ExitDispute contract that implements the DisputeInterface
・Determine the data structure of ExitClaim and ExitChallenge
・Implement claim, challenge, removeChallenge,settle

# Why

・Changing the level of abstraction of OGS

# Ref
https://cryptoeconomicslab.atlassian.net/browse/WAK-386

### CODE REVIEW

File: contracts/Dispute/ExitDispute.sol [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        );
+
+        if (keccak256(_challengeInputs[0]) == keccak256(EXIT_SPENT_CHALLENTE)) {
+            new SpentChallenge().verify(_inputs, _witness);
+        } else if (keccak256(_challengeInputs[0]) == keccak256(EXIT_CHECKPOINT_CHALLENTE)) {
+            new CheckpointChallenge().verify(_inputs, _witness);
+        } else {
+            revert("illegal challenge type");
+        }
+        disputeManager.challenge(createProperty(_inputs[0], EXIT_CLAIM), createProperty(_challengeInputs[0], _challengeInputs[0]));
```
Comment (syuhei176) [inline contracts/Dispute/ExitDispute.sol:123]: is this second argument `createProperty(_challengeInputs[0], _challengeInputs[1])` or `createProperty(_challengeInputs[0], EXIT_SPENT_CHALLENTE)`?

File: test/Dispute/ExitDispute.test.ts [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+            challengeInputs,
+            challengeWitness,
+            {
+              gasLimit: 800000
+            }
+          )
+        ).to.emit(exitDispute, 'ExitChallenged')
+      }).timeout(15000)
+    })
+    describe('failer to exit challenge', () => {
```
Comment (tkmct) [inline test/Dispute/ExitDispute.test.ts:202]: `fail to challenge an exit`?

File: test/Dispute/ExitDispute.test.ts [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+      // prepare challenge
+      const challengeInputs = [
+        Bytes.fromString('EXIT_SPENT_CHALLENGE')
+      ]
+      const challengeWitness = [
+        encodeStructable(firstBlockInfo.inclusionProof)
+      ]
+      return [inputs, challengeInputs, challengeWitness]
+    }
+    describe('succeed to exit challenge', () => {
```
Comment (tkmct) [inline test/Dispute/ExitDispute.test.ts:187]: Can you add test cases to check if right mocked challengeContract is called in succeed cases?

File: contracts/Dispute/ExitDispute.sol (hunk 1 above)
Comment (tkmct) [inline contracts/Dispute/ExitDispute.sol:123]: challenge property should be `createProperty(challengeInputs[1], _challengeInputs[0])`?

File: test/Dispute/ExitDispute.test.ts [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+    })
+  })
+  
+  describe.skip('removeChallenge', () => {
+    // not implemented
+  })
+
+  describe.skip('settle', () => {
+
+  })
```
Comment (tkmct) [inline test/Dispute/ExitDispute.test.ts:251]: Can you remove these lines?
we don't need `settle` test cases.

File: test/Dispute/ExitDispute.test.ts [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+            {
+              gasLimit: 800000
+            }
+          )
+        ).to.emit(exitDispute, 'ExitChallenged')
+      }).timeout(15000)
+    })
+    describe('failer to exit challenge', () => {
+      it('check ', async () => {
+        const [, challengeInputs, challengeWitness] = await init()
```
Comment (tkmct) [inline test/Dispute/ExitDispute.test.ts:204]: FYI: this validation rule differs for spentChallenge and checkpointChallenge.
We can add test cases when implementing challenge contracts.

File: contracts/Dispute/ExitDispute.sol (hunk 1 above)
Comment (tkmct) [reply contracts/Dispute/ExitDispute.sol:123]: I think it's also okay to create another method s.a. `createSpentChallengeProperty()` or `createCheckpointChallengeProperty()`

Comment (Akira-Taniguchi) [reply contracts/Dispute/ExitDispute.sol:123]: the length of _challengeInputs is 1 and the _challengeInputs has either EXIT_SPENT_CHALLENTE or EXIT_CHECKPOINT_CHALLENTE at the beginning of the _challengeInputs, I'm thinking that createProperty(_challengeInputs[0], _challengeInputs[0]) is fine.

@tkmct

File: test/Dispute/ExitDispute.test.ts [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+      commitmentVerifier.address,
+      utils.address
+    ])
+
+  })
+
+
+  describe('claim', () => {
+    describe('succeed to claim a checkpoint', () => {
+      it('create a new checkpoint claim', async () => {
```
[the remark this answers is not in the document]
Comment (Akira-Taniguchi) [reply test/Dispute/ExitDispute.test.ts:79]: That's right, I'll fix it.
@tkmct

File: test/Dispute/ExitDispute.test.ts (hunk 2 above)
Comment (Akira-Taniguchi) [reply test/Dispute/ExitDispute.test.ts:202]: That's right, I'll fix it.
@tkmct

File: contracts/Dispute/ExitDispute.sol (hunk 1 above)
Comment (tkmct) [reply contracts/Dispute/ExitDispute.sol:123]: length of challengeInputs is 2 when CheckpointChallenge and 1 when SpentChallenge.
Let's talk about this tomorrow.

Comment (Akira-Taniguchi) [reply contracts/Dispute/ExitDispute.sol:123]: Could it be that the length of the challengeInputs is more than 2?

@tkmct

File: test/Dispute/ExitDispute.test.ts (hunk 4 above)
Comment (Akira-Taniguchi) [reply test/Dispute/ExitDispute.test.ts:251]: ok,thanks.
I'll delete it.

@tkmct

File: contracts/Dispute/ExitDispute.sol (hunk 1 above)
Comment (tkmct) [reply contracts/Dispute/ExitDispute.sol:123]: 1 or 2.

Comment (Akira-Taniguchi) [reply contracts/Dispute/ExitDispute.sol:123]: ok, thanks

Comment (Akira-Taniguchi) [reply contracts/Dispute/ExitDispute.sol:123]: That's how it was corrected for now.

@tkmct

**Pull request merged**